resolver/resolver.go

```diff
@@ -67,147 +71,217 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (cid.
 	return nd.ResolveLink(names)
 }
 
 // ResolvePathComponents fetches the nodes for each segment of the given path.
 // It uses the first path component as a hash (key) of the first node, then
 // resolves all other components walking the links, with ResolveLinks.
-func (r *Resolver) ResolvePathComponents(ctx context.Context, fpath path.Path) ([]ipld.Node, error) {
+func (r *Resolver) ResolvePathComponents(ctx context.Context, fpath path.Path) ([]ipldp.Node, error) {
 	evt := log.EventBegin(ctx, "resolvePathComponents", logging.LoggableMap{"fpath": fpath})
 	defer evt.Done()
 
-	h, parts, err := path.SplitAbsPath(fpath)
+	// validate path
+	if err := fpath.IsValid(); err != nil {
+		return nil, err
+	}
+
+	c, p, err := path.SplitAbsPath(fpath)
 	if err != nil {
-		evt.Append(logging.LoggableMap{"error": err.Error()})
 		return nil, err
 	}
 
-	log.Debug("resolve dag get")
-	nd, err := r.DAG.Get(ctx, h)
+	// create a selector to traverse all path segments but only match the last
+	pathSelector, err := pathAllSelector(p)
 	if err != nil {
-		evt.Append(logging.LoggableMap{"error": err.Error()})
 		return nil, err
 	}
 
-	return r.ResolveLinks(ctx, nd, parts)
+	nodes, _, _, err := r.resolveNodes(ctx, c, pathSelector)
+	return nodes, err
 }
 
 // ResolveLinks iteratively resolves names by walking the link hierarchy.
-// Every node is fetched from the DAGService, resolving the next name.
+// Every node is fetched from the fetcher, resolving the next name.
 // Returns the list of nodes forming the path, starting with ndd. This list is
 // guaranteed never to be empty.
 //
 // ResolveLinks(nd, []string{"foo", "bar", "baz"})
 // would retrieve "baz" in ("bar" in ("foo" in nd.Links).Links).Links
-func (r *Resolver) ResolveLinks(ctx context.Context, ndd ipld.Node, names []string) ([]ipld.Node, error) {
+func (r *Resolver) ResolveLinks(ctx context.Context, ndd ipldp.Node, names []string) ([]ipldp.Node, error) {
 
 	evt := log.EventBegin(ctx, "resolveLinks", logging.LoggableMap{"names": names})
 	defer evt.Done()
-	result := make([]ipld.Node, 0, len(names)+1)
-	result = append(result, ndd)
-	nd := ndd // dup arg workaround
-
-	// for each of the path components
-	for len(names) > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, time.Minute)
-		defer cancel()
-
-		lnk, rest, err := r.ResolveOnce(ctx, r.DAG, nd, names)
-		if err == dag.ErrLinkNotFound {
-			evt.Append(logging.LoggableMap{"error": err.Error()})
-			return result, ErrNoLink{Name: names[0], Node: nd.Cid()}
-		} else if err != nil {
-			evt.Append(logging.LoggableMap{"error": err.Error()})
-			return result, err
+
+	// create a selector to traverse all path segments but only match the last
+	pathSelector, err := pathAllSelector(names)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a new cancellable session
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
```

Comment:
@Stebalien any idea if this is still necessary? Tracking which of these things will internally timeout vs not is pretty confusing especially when the functions already take a context.

Reply:
see comment above -- I think we should preserve behavior where possible.

Reply:
and remove in a future release.

Reply:
Removing in a future release sounds reasonable (easier to track down bugs)